contract/models/abstract_contract.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2004-2010 OpenERP SA
+# Copyright 2014 Angel Moya <angel.moya@domatix.com>
+# Copyright 2015-2020 Tecnativa - Pedro M. Baeza
+# Copyright 2016-2018 Carlos Dauden <carlos.dauden@tecnativa.com>
+# Copyright 2016-2017 LasLabs Inc.
+# Copyright 2018 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class ContractAbstractContract(models.AbstractModel):
+    _inherit = "contract.recurrency.basic.mixin"
+    _name = "contract.abstract.contract"
+    _description = "Abstract Recurring Contract"
+
+    # These fields will not be synced to the contract
+    NO_SYNC = ["name", "partner_id", "company_id"]
+
+    name = fields.Char(required=True)
+    # Needed for avoiding errors on several inherited behaviors
+    partner_id = fields.Many2one(
+        comodel_name="res.partner", string="Partner", index=True
+    )
+    pricelist_id = fields.Many2one(comodel_name="product.pricelist", string="Pricelist")
+    contract_type = fields.Selection(
+        selection=[("sale", "Customer"), ("purchase", "Supplier")],
+        default="sale",
+        index=True,
+    )
+
+    journal_id = fields.Many2one(
+        "account.journal",
+        string="Journal",
+        default=lambda s: s._default_journal(),
+        domain="[('type', '=', contract_type)," "('company_id', '=', company_id)]",
+        index=True,
+    )
+    company_id = fields.Many2one(
+        "res.company",
+        string="Company",
+        required=True,
+        default=lambda self: self.env.company.id,
+    )
+    line_recurrency = fields.Boolean(
+        string="Recurrency at line level?",
+        help="Mark this check if you want to control recurrrency at line level instead"
+        " of all together for the whole contract.",
+    )
+
+    @api.onchange("contract_type")
+    def _onchange_contract_type(self):
+        if self.contract_type == "purchase":
+            self.contract_line_ids.filtered("automatic_price").update(
+                {"automatic_price": False}
+            )
+        self.journal_id = self.env["account.journal"].search(
+            [
+                ("type", "=", self.contract_type),
+                ("company_id", "=", self.company_id.id),
+            ],
+            limit=1,
+        )
+
+    @api.model
+    def _default_journal(self):
+        company_id = self.env.context.get("company_id", self.env.user.company_id.id)
+        domain = [
+            ("type", "=", self.contract_type),
+            ("company_id", "=", company_id),
+        ]
+        return self.env["account.journal"].search(domain, limit=1)
```

Comment:
When creating contract using `create` with given "contract_type" value, self.contract_type is false. So default journal is not set.
Anyway it seems like Odoo's bug. At least Odoo 13.0-20200417

Reply:
`default_contract_type` is set in context from the menu action. `self.contract_type` is not false if the contract is created from the UI.

Need to be explicitly added to context if the contract creation is scripted.

Reply:
default_contract_type is set in context from the menu action. self.contract_type is not false if the contract is created from the UI.

Need to be explicitly added to context if the contract creation is scripted.

Reply:
I'm replacing current code with a computed writable field that should get rid of this problem. Please check when pushed.